### DISCUSSION TITLE
Thoughts about Anvil?

### DISCUSSION THREAD

**Discussion by hugetim** [AnswerDotAI/fasthtml#199] [General]
For someone who likes using Python and doesn't yet have much knowledge of the HTML, CSS, and JavaScript that using FastHTML seems to require, how would you compare the pros and cons of using [Anvil](https://anvil.works/) for a new web app verses FastHTML?

**Comment by bryanwhiting** [comment]
i've never used anvil, though I've looked into it, and have barely used fasthtml. 

anvil solves a lot of problems for you. if you want a quick web app up and running, anvil will be quick and fasthtml is like coding a website from raw HTML. so anvil will hide a lot of stuff from you and limit you from being able to do *whatever* you want (if you want to be able to do *whatever*), and fasthtml will allow you to build...*whatever* you want.

fasthtml is unique compared to streamlit and gradio in that it's going to be foundational. Like, you'll be able to build any website you want in it. streamlit/gradioare focused on simplifying the experience to get you going quickly. But if you change how it looks you're pretty constrained. anvil seems more flexible. I can see fasthtml getting to a point in the next 6 months where there are templates that will simplify a lot of this, perhaps like shinydashboard. Right now, in its early days, you do need to know html + css to something that looks decent.

anvil also has some cool things where it connects to your data model, etc. I personally don't see fasthtml going that direction. So, i think you're going to be able to build one-person apps much quicker in anvil, but real production-grade web apps will be built using fasthtml by big companies in a few years. anvil will probably cap out at being a customer-facing app, and fasthtml will be a competitor/replacement for things like nextjs.

**Reply by hugetim** [reply]
Right, well, as you seem to have partially realized when editing your answer, Anvil is quite different from streamlit and gradio. I'm not sure what you have in mind when saying it hides a lot of stuff from you. It's pretty much all there, in my understanding, accessible via [various](https://anvil.works/docs/ui/custom-styling) [escape](https://anvil.works/docs/client/javascript) [hatches](https://anvil.works/docs/data-tables/external-database). My understanding of Anvil's limitations: generally single-page applications only (though with some [signs of change on that](https://anvil.works/forum/t/serve-anything-anywhere-on-your-domain-with-anvil-server-route/20919)), which slows initial loading given the volume of JavaScript required, and lower-priced tiers of the hosted option have [slow server calls](https://anvil.works/docs/server#persistent-server-modules) (and the open-source self-hosted option seems to also involve performance compromises). 

My sense is that Anvil has gone beyond customer-facing apps in [some cases](https://anvil.works/for/new-products), but you are right that that seems to be the main use case beyond hobby apps (like my own).

I guess I will probably stick with Anvil for my next app, but I'm excited to see where this goes.
